tools/release/stage-release.ts

```diff
@@ -0,0 +1,182 @@
+import {bold, cyan, green, italic, red, yellow} from 'chalk';
+import {existsSync, readFileSync, writeFileSync} from 'fs';
+import {prompt} from 'inquirer';
+import {join} from 'path';
+import {GitClient} from './git/git-client';
+import {promptForNewVersion} from './prompt/new-version-prompt';
+import {parseVersionName, Version} from './version-name/parse-version';
+import {getExpectedPublishBranch} from './version-name/publish-branch';
+
+/**
+ * Class that can be instantiated in order to stage a new release. The tasks requires user
+ * interaction/input through command line prompts.
+ *
+ * Staging a release involves the following the steps:
+ *
+ *  1) Prompt for release type (with version suggestion)
+ *  2) Prompt for version name if no suggestions has been selected
+ *  3) Assert that the proper publish branch is checked out (e.g. 6.4.x for patches)
+ *  4) Assert that there are no local changes which are uncommitted.
+ *  5) Assert that the local branch is up to date with the remote branch.
+ *  6) Creates a new branch for the release staging (release-stage/{VERSION})
+ *  7) Switches to the staging branch and updates the package.json
+ *  8) Waits for the user to continue (users can generate the changelog in the meanwhile)
+ *  9) Create a commit that includes all changes in the staging branch.
+ */
+class StageReleaseTask {
+
+  /** Path to the project package JSON. */
+  packageJsonPath: string;
+
+  /** Serialized package.json of the specified project. */
+  packageJson: any;
+
+  /** Parsed current version of the project. */
+  currentVersion: Version;
+
+  /** Instance of a wrapper that can execute Git commands. */
+  git: GitClient;
+
+  constructor(public projectDir: string) {
+    this.packageJsonPath = join(projectDir, 'package.json');
+
+    if (!existsSync(this.packageJsonPath)) {
+      console.error(red(`The specified directory is not referring to a project directory.` +
+        `There must be a ${italic('package.json')} file in the project directory.`));
+      process.exit(1);
+    }
+
+    this.packageJson = JSON.parse(readFileSync(this.packageJsonPath, 'utf-8'));
+    this.currentVersion = parseVersionName(this.packageJson.version);
+
+    if (!this.currentVersion) {
+      console.error(red(`Cannot parse current version in ${italic('package.json')}. Please ` +
+        `make sure "${this.packageJson.version}" is a valid Semver version.`));
+      process.exit(1);
+    }
+
+    this.git = new GitClient(projectDir, this.packageJson.repository.url);
+  }
+
+  async run() {
+    console.log();
+    console.log(cyan('-----------------------------------------'));
+    console.log(cyan('  Angular Material stage release script'));
+    console.log(cyan('-----------------------------------------'));
+    console.log();
+
+    const newVersion = await promptForNewVersion(this.currentVersion);
```

Comment:
Is there anything that validates this new version? Mainly for the case where a custom version is entered.

Reply:
Yeah, the custom version will be already validated by the `validate` function which you've simplified with the comment above.

Reply:
I meant more that the version selected is the correct version, not that it's _a_ valid version. E.g., entering `1.0.0` being incorrect. 

This could go in a follow-up, though.

Reply:
Ah I see, so a custom version can be any valid Semver version that is more recent than the current version right?

Otherwise, if we add more limitations, we could probably remove the "custom-release" once there is a way to create pre-releases with a label

Reply:
Yeah, removing it might be better because I can't think of a scenario where you'd want to do something abnormal.

Reply:
I agree. I will remove it in the follow-up that allows creating a pre-release with custom label.